### PR TITLE
Add Nix support for building F*

### PR DESCRIPTION
**Pull request by W95Psp**
Hi,

@pnmadelaine and I have been experimenting building F* with the [Nix package manager](https://nixos.org/).
Nix allows for very high guarantees concerning reproducibility: we believe having a Nix build right inside the F* repository would be a nice thing.

This PR makes the changes listed below.
 - Add a `flake.nix` file, which turns the F\* Git repository into a [Nix flake](https://nixos.wiki/wiki/Flakes), that allows Nix users to build and run F\* at any commit/branch/reference/pr (e.g. `nix run github:FStarLang/FStar/my-branch` will build & run F* at branch `my-branch`).
 - Add a `flake.lock` file (similar to Cargo's `Cargo.lock` or to NPM's `package-lock.json`) that locks every inputs declared in the Nix file `flake.nix`.
 - Add a `.nix/lib.nix` file, that contains different Nix functions that realize the bootstrap process of F* (so that Nix users can choose whether they want to bootstrap or not, and how), build the F* book, etc.
 - Add a `.nix/README.md` file, that explains how to use the "fine-grained" Nix API for building F*.
 - Add a `Nix Package` section in `INSTALL.md`.

For some time we kept the Nix build in the [hacl-nix](https://github.com/hacl-star/hacl-nix) repository, but:
 - this is poorly discoverable;
 - this is tied to Hacl*;
 - Nix makes it really easy to build one specific commit in the commit history when the Nix files live alongside the sources, for instance, one can search dichotomously for a regression very easily.

### CODE REVIEW

File: .nix/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+with `github:repo/user`.
+
+*Note that running any of the command above will trigger (if not
+cached locally) a build of F\* from sources, which takes some time.*
+
+## Bootstraping
+
+The `flake.nix` at the root of this repository builds the current
+OCaml snapshot by default (package named `fstar`), but also exposes
+the full bootsrap build via a package named `fstar-bootsrap`.
```
Comment (nikswamy) [inline .nix/README.md:40]: typos: bootsrap and fstar-bootsrap

**Comment by nikswamy** [comment]
I tried out the instructions for the basic scenario of getting a nix shell with fstar and it worked beautifully!

File: .nix/README.md (hunk 1 above)
Comment (W95Psp) [reply .nix/README.md:40]: Thanks!

**Comment by W95Psp** [comment]
Thanks for the feedback, I'm glad it went smoothly :)

**Pull request merged**